inspirehep/modules/orcid/tasks.py

```diff
@@ -93,35 +89,63 @@ def _link_user_and_token(user, name, orcid, token):
         # User already has their ORCID linked
         pass
 
-    # Check whether there are already tokens associated with this
-    # ORCID identifier.
-    tokens = RemoteToken.query.join(RemoteAccount).join(User)\
-        .join(UserIdentity).filter(UserIdentity.id == orcid).all()
+    # Search for existent tokens.
+    # Note: there can be only 1 RemoteToken per given RemoteAccount.
+    existent_remote_token = RemoteToken.query.join(RemoteAccount).filter_by(
+        user=user).one_or_none()
 
-    if tokens:
-        # Force the allow_push.
-        with db.session.begin_nested():
-            for token in tokens:
-                if not token.remote_account.extra_data['allow_push']:
-                    result = orcid
-                token.remote_account.extra_data['allow_push'] = True
-    else:
-        # If not, create and put the token entry
+    # If not existent_remote_token, create one.
+    if not existent_remote_token:
         with db.session.begin_nested():
-            result = orcid
             RemoteToken.create(
                 user_id=user.id,
                 client_id=get_value(current_app.config, 'ORCID_APP_CREDENTIALS.consumer_key'),
-                token=token,
+                token=access_token,
                 secret=None,
                 extra_data={
                     'orcid': orcid,
                     'full_name': name,
                     'allow_push': True,
                 }
             )
-
-    return result
+        return orcid
+
+    # If there is an existent_remote_token:
+    #    ensure it is associated to this ORCID and
+    #    set allow_push to True.
+    #
+    # Get the ORCID associated with this token.
+    user_identity = UserIdentity.query.filter_by(user=user, method='orcid').one_or_none()
+
+    # Ensure data consistency.
+    if not user_identity:
+        raise Exception('No UserIdentity for user={}, method="orcid", while'
```

Comment:
this will raise for all tokens migrated for legacy, as `UserIdentity` was never created for them. Is that done on purpose?